SpokeStack/TFLiteWakewordRecognizer.swift

```diff
@@ -0,0 +1,478 @@
+            /// Compute the stft spectrogram
+            self.fft.forward(&self.fftFrame)
+            
+            /// rewind the sample window for another run
+            self.sampleWindow.rewind().seek(self.hopLength)
+            
+            //self.fftFrameCollector += "\(self.fftFrame)\n"
+        }
+    }
+    
+    /// Attention model processing
+    
+    private func filter() throws -> Void {
+        if !self.fftFrame.isEmpty {
+            do {
+                if let model = self.filterModel {
+                    /// inputs
+                    /// compute the manitude of the spectrogram
+                    let magnitude = (self.fftFrame.count / 2) + 1
+                    /// copy the spectrogram into the filter model's input
+                    _ = try self
+                        .fftFrame
+                        .prefix(magnitude)
+                        .withUnsafeBytes(
+                            {try model.copy(Data($0), toInputAt: 0)})
+                    
+                    /// calculate
+                    try model.invoke()
+                    
+                    /// outputs
+                    let output = try model.output(at: 0)
+                    let results = output.data.withUnsafeBytes { (pointer: UnsafePointer<Float32>) -> [Float32] in
+                        Array<Float32>(UnsafeBufferPointer(start: pointer, count: output.data.count / 4))}
+                    self.frameWindow.rewind().seek(self.melWidth)
+                    for r in results {
+                        try self.frameWindow.write(r)
+                        //filterCollector.append(r)
+                    }
+                } else {
+                    throw WakewordModelError.filter("model is not initialized")
+                }
+                
+            } catch let message {
+                throw WakewordModelError.filter("TFLiteWakewordRecognizer filter \(message)")
+            }
+        }
+    }
+    
+    private func encode() throws -> Void {
+        if self.frameWindow.isFull {
+            do {
+                if let model = self.encodeModel {
+                    /// inputs
+                    self.frameWindow.rewind()
+                    // TODO: model.copy requires that the data be sized to exactly the same as the tensor, so we can't just do read()s off the ringbuffer and copy over piecewise. This introduces an aggrevating overhead of having to copy the ringbuffer into an array before copying over to the tensor. Maybe use a fixed-sized array that is advanced based off the fft frame size?
+                    var frameWindowArray: Array<Float32> = []
+                    while !self.frameWindow.isEmpty {
+                        let f = try self.frameWindow.read()
+                        frameWindowArray.append(f)
+                    }
+                    var stateArray: Array<Float32> = []
+                    for _ in 0..<self.stateWidth {
+                        let f = try self.encodeState.read()
```

Comment:
I was thinking that somebody would need to reset the `encodeState` ring buffer, but it is neatly reset every time it is read/written. Nice.